save-orchestrator/src/test/kotlin/org/cqfn/save/orchestrator/controller/heartbeat/HeartbeatControllerTest.kt

```diff
@@ -73,6 +93,106 @@ class HeartbeatControllerTest {
         assertTrue(monoResponse.ids.isNotEmpty() && monoResponse.ids.first().filePath == "qwe")
     }
 
+    @Test
+    fun `should not shutdown any agents when not all of them are IDLE`() {
+        testHeartbeat(
+            agentStatuses = listOf(
+                AgentStatus(LocalDateTime.now(), AgentState.IDLE, Agent("test-1", null)),
+                AgentStatus(LocalDateTime.now(), AgentState.BUSY, Agent("test-2", null)),
+            ),
+            heartbeat = Heartbeat("test-1", AgentState.IDLE, ExecutionProgress(100)),
+            tests = emptyList(),
+            mockAgentStatuses = true,
+        ) {
+            verify(dockerService, times(0)).stopAgents(any())
+        }
+    }
+
+    @Test
+    fun `should not shutdown any agents when all agents are IDLE but there are more tests left`() {
+        testHeartbeat(
+            agentStatuses = listOf(
+                AgentStatus(LocalDateTime.now(), AgentState.IDLE, Agent("test-1", null)),
+                AgentStatus(LocalDateTime.now(), AgentState.IDLE, Agent("test-2", null)),
+            ),
+            heartbeat = Heartbeat("test-1", AgentState.IDLE, ExecutionProgress(100)),
+            tests = listOf(
+                TestDto("/path/to/test-1", 1, 1),
+                TestDto("/path/to/test-2", 1, 2),
+                TestDto("/path/to/test-3", 1, 3),
+            ),
+            mockAgentStatuses = false,
+        ) {
+            verify(dockerService, times(0)).stopAgents(any())
+        }
+    }
+
+    @Test
+    fun `should shutdown idle agents when there are no tests left`() {
+        testHeartbeat(
+            agentStatuses = listOf(
+                AgentStatus(LocalDateTime.now(), AgentState.IDLE, Agent("test-1", null)),
+                AgentStatus(LocalDateTime.now(), AgentState.IDLE, Agent("test-2", null)),
+            ),
+            heartbeat = Heartbeat("test-1", AgentState.IDLE, ExecutionProgress(100)),
+            tests = emptyList(),
+            mockAgentStatuses = true,
+        ) {
+            verify(dockerService, times(1)).stopAgents(any())
+        }
+    }
+
+    /**
+     * Test logic triggered by a heartbeat.
+     *
+     * @param agentStatuses agent statuses that are returned from backend (mocked response)
+     * @param heartbeat a [Heartbeat] that is received by orchestrator
+     * @param tests a batch of tests returned from backend (mocked response)
+     * @param mockAgentStatuses whether a mocked response for `/getAgentsStatusesForSameExecution` should be added to queue
+     * @param verification a lambda for test assertions
+     */
+    private fun testHeartbeat(
+        agentStatuses: List<AgentStatus>,
+        heartbeat: Heartbeat,
+        tests: List<TestDto>,
+        mockAgentStatuses: Boolean = false,
+        verification: () -> Unit,
```

Comment:
Do you need to change it to a _agentStatuses: List<AgentStatusDto>_?

Reply:
Yes, updated